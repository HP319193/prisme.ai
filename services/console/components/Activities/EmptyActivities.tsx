import { Trans, useTranslation } from 'next-i18next';
import { useWorkspaceLayout } from '../../layouts/WorkspaceLayout/context';
import SuggestionButton from '../SuggestionButton';

export const EmptyActivities = () => {
  const { t } = useTranslation('workspaces');
  const { createAutomation, createPage, installApp } = useWorkspaceLayout();
  return (
    <div className="flex flex-1 justify-between flex-col p-28">
      <div className="flex flex-col flex-1 justify-center w-[17rem]">
        <Trans t={t} i18nKey="workspace.empty.title">
          <div className="font-bold text-2xl my-2">a</div>
          <div>b</div>
        </Trans>
      </div>
      <div className="flex flex-col">
        <div className="my-8">{t('workspace.empty.subtitle')}</div>
        <div className="flex flex-row">
          <SuggestionButton
            title={t('workspace.empty.suggestions.page.title')}
            text={t('workspace.empty.suggestions.page.text')}
            className="!ml-0"
            color="#F7F8D0"
            onClick={() => createPage()}
          />
          <SuggestionButton
            title={t('workspace.empty.suggestions.automation.title')}
            text={t('workspace.empty.suggestions.automation.text')}
            color="#FF9261"
            onClick={createAutomation}
          />
          <SuggestionButton
            title={t('workspace.empty.suggestions.app.title')}
            text={t('workspace.empty.suggestions.app.text')}
            color="#E7F6F6"
            onClick={installApp}
          />
        </div>
      </div>
    </div>
  );
};

export default EmptyActivities;
