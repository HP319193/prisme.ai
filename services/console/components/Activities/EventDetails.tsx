import { FC, MouseEvent, useCallback, useMemo } from 'react';
import { Table, Tooltip } from '@prisme.ai/design-system';
import { Event } from '@prisme.ai/sdk';
import { useTranslation } from 'next-i18next';
import { selectText } from '../../utils/dom';
import { truncate } from '../../utils/strings';
import {
  FilterOutlined,
  PlayCircleOutlined,
  PlusCircleOutlined,
} from '@ant-design/icons';
import { useWorkspace } from '../../providers/Workspace';
import { useQueryString } from '../../providers/QueryStringProvider';

interface EventsDetailsProps {
  event: Event<Date>;
}

const PAYLOAD_TRUNCATE_LENGTH = 400;

const truncatePayload = (payloadValue: string) => {
  if (payloadValue && payloadValue.length > PAYLOAD_TRUNCATE_LENGTH) {
    return truncate(payloadValue, PAYLOAD_TRUNCATE_LENGTH);
  }
  return payloadValue;
};

interface EventRecord {
  key: string;
  name: string;
  value: any;
  payloadValue: any;
  fullPayload: string;
}

export const EventDetails: FC<EventsDetailsProps & { replay: () => void }> = ({
  replay,
  event,
}) => {
  const { t } = useTranslation('workspaces');
  const { setQueryString } = useQueryString();
  const dataSource = useMemo(() => {
    const stringifiedPayload = JSON.stringify(event.payload, null, ' ');
    return [
      {
        key: 'type',
        name: 'type',
        value: event.type,
      },
      event.source.appInstanceFullSlug
        ? {
            key: 'source.appInstanceFullSlug',
            name: 'source.appInstanceFullSlug',
            value: event.source.appInstanceFullSlug,
          }
        : false,
      event.source.automationSlug
        ? {
            key: 'source.automationSlug',
            name: 'source.automationSlug',
            value: event.source.automationSlug,
          }
        : false,
      {
        key: 'payload',
        name: 'payload',
        value: (
          <pre>
            <code>{truncatePayload(stringifiedPayload)}</code>
          </pre>
        ),
        payloadValue: event.payload && (
          <pre>
            <code>{stringifiedPayload}</code>
          </pre>
        ),
        fullPayload: stringifiedPayload,
      },
      {
        key: 'source.userId',
        name: 'source.userId',
        value: event.source.userId,
      },
      {
        key: 'source.sessionId',
        name: 'source.sessionId',
        value: event.source.sessionId,
      },
      {
        key: 'source.correlationId',
        name: 'source.correlationId',
        value: event.source.correlationId,
      },
      event.error
        ? {
            key: 'error.error',
            name: 'error.error',
            value: event.error?.error,
          }
        : false,
      event.error
        ? {
            key: 'error.message',
            name: 'error.message',
            value: event.error?.message,
          }
        : false,
      event.error
        ? {
            key: 'error.details',
            name: 'error.details',
            value: event.error?.details && (
              <pre>
                <code>{JSON.stringify(event.error?.details, null, ' ')}</code>
              </pre>
            ),
          }
        : false,
      event.target
        ? {
            key: 'target',
            name: 'target',
            value: event.target && (
              <pre>
                <code>{JSON.stringify(event.target, null, ' ')}</code>
              </pre>
            ),
          }
        : false,
      {
        key: 'id',
        name: 'id',
        value: event.id,
      },
      {
        key: 'size',
        name: 'size',
        value: event.size,
      },
    ].filter(Boolean) as EventRecord[];
  }, [event]);
  const onRowClick = useCallback(({ target }: MouseEvent) => {
    const valueTd = (
      target as HTMLTableRowElement
    ).parentNode?.querySelectorAll('td');
    if (valueTd && valueTd.length === 3) {
      selectText(valueTd[2]);
    }
  }, []);

  const isEmit = event?.source?.serviceTopic === 'topic:runtime:emit';

  const addFilter = useCallback(
    (k: string, v: string) => {
      setQueryString((prev) => {
        const newQuery = new URLSearchParams(prev);
        newQuery.set(k, v);
        return newQuery;
      });
    },
    [setQueryString]
  );

  return (
    <div className="relative">
      <Table
        dataSource={dataSource}
        columns={[
          { title: t('events.details.name'), dataIndex: 'name', key: 'name' },
          {
            title: t('events.details.value'),
            dataIndex: 'value',
            key: 'value',
            render: (children, data) => {
              if (typeof children === 'object') return children;
              return (
                <>
                  {children}
                  <Tooltip title={t('events.details.addFilter')}>
                    <button
                      className="relative text-gray ml-2"
                      onClick={() => addFilter(data.name, data.value)}
                    >
                      <FilterOutlined />
                      <PlusCircleOutlined className="absolute bottom-0 -right-1 text-xs" />
                    </button>
                  </Tooltip>
                </>
              );
            },
          },
        ]}
        bordered
        pagination={{ pageSize: 50, position: [] }}
        scroll={{ y: 500 }}
        expandable={{
          expandedRowRender: (record) => record.payloadValue,
          rowExpandable: (record) =>
            record.name === 'payload' &&
            !!record.fullPayload &&
            record.fullPayload.length > PAYLOAD_TRUNCATE_LENGTH,
        }}
        expandRowByClick
        onRow={() => ({
          onClick: onRowClick,
        })}
      />
      {isEmit && (
        <Tooltip title={t('feed.replay')} placement="right">
          <button
            className="absolute top-[1.3rem] left-[1.3rem]"
            onClick={replay}
          >
            <PlayCircleOutlined />
          </button>
        </Tooltip>
      )}
    </div>
  );
};

export const LinkedEventDetails: FC<Event<Date>> = (event) => {
  const { events } = useWorkspace();
  const replay = useCallback(() => {
    events.emit(event.type, event.payload);
  }, [events, event.payload, event.type]);
  return useMemo(
    () => <EventDetails event={event} replay={replay} />,
    [event, replay]
  );
};

export default LinkedEventDetails;
