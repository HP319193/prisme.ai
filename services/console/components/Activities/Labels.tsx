import { WarningOutlined } from '@ant-design/icons';
import { Button, notification } from '@prisme.ai/design-system';
import { Popconfirm } from 'antd';
import { useTranslation } from 'next-i18next';
import Link from 'next/link';
import { FC } from 'react';
import { useWorkspace } from '../../providers/Workspace';
import api, { Event } from '../../utils/api';

export const AutomationLabel: FC<Event<Date>> = ({
  children,
  source,
  payload,
}) => {
  const { workspace } = useWorkspace();

  if (source.appSlug) {
    return <span className="font-bold">{children}</span>;
  }
  const automationSlug =
    payload.slug || payload.automationSlug || source.automationSlug;
  const automationExists = !!workspace?.automations?.[automationSlug];

  if (automationExists) {
    return (
      <Link
        href={`/workspaces/${source.workspaceId}/automations/${automationSlug}`}
      >
        <a className="font-bold">{children}</a>
      </Link>
    );
  }
  return <span className="font-bold">{children}</span>;
};

export const PageLabel: FC<Event<Date>> = ({ children, payload, source }) => {
  return (
    <Link href={`/workspaces/${source.workspaceId}/pages/${payload.page.id}`}>
      <a className="font-bold">{children}</a>
    </Link>
  );
};

export const AppLabel: FC<Event<Date>> = ({ children, payload, source }) => {
  return (
    <Link
      href={`/workspaces/${source.workspaceId}/apps/${payload?.appInstance?.appName}`}
    >
      <a className="font-bold">{children}</a>
    </Link>
  );
};

export const ErrorLabel: FC<Event<Date>> = ({ children, type }) => {
  return (
    <div className="flex items-center">
      <span className="flex mr-[0.5rem] text-pr-orange">
        <WarningOutlined />
      </span>
      <span className="flex">{children}</span>
    </div>
  );
};

export const EventLabel: FC<Event<Date>> = ({ children }) => {
  return <span className="font-bold">{children}</span>;
};

export const RollbackVersion: FC<Event<Date>> = ({
  children,
  source: { workspaceId },
  payload: { version },
}) => {
  const { t } = useTranslation('workspaces');
  const { fetchWorkspace } = useWorkspace();
  return (
    <Popconfirm
      title={t('workspace.versions.rollback.confirm')}
      okText={t('workspace.versions.rollback.confirm_ok')}
      cancelText={t('cancel', { ns: 'common' })}
      onCancel={(e) => e?.stopPropagation()}
      onConfirm={async (e) => {
        e?.stopPropagation();
        if (!workspaceId || !version) return;
        await api.workspaces(workspaceId).versions.rollback(version.name, {
          repository: version.repository,
        });
        fetchWorkspace();
        notification.success({
          message: t('workspace.versions.rollback.success'),
          placement: 'bottomRight',
        });
      }}
    >
      <Button className="font-bold" onClick={(e) => e.stopPropagation()}>
        {children}
      </Button>
    </Popconfirm>
  );
};
