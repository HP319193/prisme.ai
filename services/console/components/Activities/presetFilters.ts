export const filters = ({ sessionId }: { sessionId?: string }) => ({
  versions: {
    type: 'workspaces.versions.published',
  },
  errors: {
    types: 'error,runtime.fetch.failed',
  },
  updates: {
    type: 'workspaces.*',
  },
  webhooks: {
    type: 'runtime.interactions.triggered',
    'payload.trigger.type': 'endpoint',
  },
  shares: {
    type: 'workspaces*permissions.*',
  },
  application: {
    'source.serviceTopic': 'topic:runtime:emit',
  },
  inputEvents: {
    'source.serviceTopic': 'topic:runtime:emit',
    'source.userId': '*',
  },
  mySession: {
    'source.sessionId': `${sessionId}`,
    'source.socketId': '*',
  },
  interactions: {
    type: 'runtime.interactions.triggered',
  },
});
