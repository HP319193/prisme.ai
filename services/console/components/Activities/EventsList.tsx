import { Collapse, Feed } from '@prisme.ai/design-system';
import { useMemo } from 'react';
import { EventsContext } from '../../providers/Events';
import { Event } from '../../utils/api';
import { useDateFormat } from '../../utils/dates';
import useLocalizedText from '../../utils/useLocalizedText';
import SourceDetails from '../SourceDetails';
import EventDetails from './EventDetails';
import SectionContent from './SectionContent';

interface EventsListProps {
  events: Set<Event<Date>>;
  read: EventsContext['read'];
  isRead: EventsContext['isRead'];
  workspaceName: string;
}

export const EventLabel = ({
  read,
  workspaceName,
  ...event
}: Event<Date> & { read: boolean; workspaceName: string }) => {
  const { localize } = useLocalizedText();
  const dateFormat = useDateFormat();

  return (
    <SourceDetails
      workspaceId={event.source.workspaceId}
      appSlug={event.source.appSlug}
    >
      <SectionContent
        title={event.source.appSlug || localize(workspaceName)}
        date={dateFormat(event.createdAt, {
          relative: true,
        })}
        type={event.type}
        read={read}
        event={event}
      />
    </SourceDetails>
  );
};

export const EventsList = ({
  events,
  workspaceName,
  read,
  isRead,
}: EventsListProps) => {
  const dateFormat = useDateFormat();

  const sections = useMemo(() => {
    const byDate = Array.from(events).reduce<[string, Set<Event<Date>>][]>(
      (prev, event) => {
        const eventDay =
          dateFormat(new Date(event.createdAt), {
            relative: true,
            withoutHour: true,
          }) || '';
        let prevDate = prev.find(([date]) => date === eventDay);
        if (!prevDate) {
          prevDate = [eventDay, new Set()];
          prev.push(prevDate);
        }
        prevDate[1].add(event);
        return prev;
      },
      []
    );

    return byDate.map(([title, events]) => {
      return {
        key: title,
        title,
        content: (
          <Collapse
            items={Array.from(events).map((event) => ({
              key: event.id,
              label: (
                <EventLabel
                  {...event}
                  read={isRead(event.id)}
                  workspaceName={workspaceName}
                />
              ),
              content: <EventDetails {...event} />,
            }))}
            light
            onChange={(ids) => {
              const id = ids[ids.length - 1];
              if (!id) return;
              read(id);
            }}
          />
        ),
      };
    });
  }, [dateFormat, events, isRead, read, workspaceName]);

  return <Feed sections={sections} />;
};

export default EventsList;
