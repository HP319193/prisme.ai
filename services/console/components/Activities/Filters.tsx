import { Button, DatePicker } from '@prisme.ai/design-system';
import { useTranslation } from 'react-i18next';
import { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { Field, Form, FormSpy } from 'react-final-form';
import { filters } from './presetFilters';
import ObjectInput from '../ObjectInput';
import moment from 'moment';
import { useQueryString } from '../../providers/QueryStringProvider';
import { Popover } from 'antd';
import { FilterOutlined } from '@ant-design/icons';

type FiltersValue = {
  beforeDate?: moment.Moment;
  afterDate?: moment.Moment;
  fields?: Record<string, string>;
};

function getInitialValue(prevQuery: URLSearchParams) {
  const beforeDate = prevQuery.get('beforeDate');
  const afterDate = prevQuery.get('afterDate');
  const value: FiltersValue = {
    beforeDate: beforeDate ? moment(beforeDate) : undefined,
    afterDate: afterDate ? moment(afterDate) : undefined,
    fields: Array.from(prevQuery.keys()).reduce(
      (prev, k) =>
        ['beforeDate', 'afterDate', 'text'].includes(k)
          ? prev
          : {
              ...prev,
              [k]: prevQuery.get(k),
            },
      {}
    ),
  };
  return value;
}

const Filters = () => {
  const { t } = useTranslation('workspaces');
  const { queryString, setQueryString } = useQueryString();
  const [suggestionsPopupState, setSuggestionsPopupState] = useState(false);
  const fieldsInput = useRef<HTMLDivElement>(null);

  const updateValue = useCallback(
    ({ values }: { values: FiltersValue }) => {
      setQueryString((prevQuery) => {
        const newQuery = new URLSearchParams();
        const text = prevQuery.get('text');
        if (text) {
          newQuery.set('text', text);
        }
        if (values.beforeDate) {
          newQuery.set('beforeDate', values.beforeDate.format());
        }
        if (values.afterDate) {
          newQuery.set('afterDate', values.afterDate.format());
        }
        Object.entries(values.fields || {}).forEach(([k, v]) => {
          newQuery.set(k, v);
        });
        if (newQuery.toString() === prevQuery.toString()) return prevQuery;
        return newQuery;
      });
    },
    [setQueryString]
  );

  const reset = useCallback(() => {
    updateValue({ values: {} });
  }, [updateValue]);

  const builtinFilters = useMemo(
    () =>
      Object.entries(filters).map(([key, filter]) => ({
        value: key,
        label: t('events.filters.suggestions.label', { context: key }),
        filter,
      })),
    [t]
  );

  const focusFieldsInput = useCallback(() => {
    setTimeout(() => {
      if (!fieldsInput.current) return;
      const [, input] = Array.from(
        fieldsInput.current.querySelectorAll('input')
      ).reverse();
      if (!input) return;
      input.focus();
    }, 200);
  }, []);

  return (
    <div className="flex flex-1 px-8">
      <Form onSubmit={updateValue} initialValues={getInitialValue(queryString)}>
        {() => (
          <form className="flex flex-1 flex-col">
            <FormSpy onChange={updateValue}></FormSpy>
            <div className="flex flex-row flex-1 justify-between py-4">
              <div className="flex flex-row">
                <div className="flex mr-4">
                  <Field name="afterDate">
                    {({ input }) => (
                      <label className="flex flex-row items-center">
                        <span className="mr-2">
                          {t('events.filters.afterDate')}
                        </span>
                        <DatePicker
                          showNow={true}
                          showTime={{
                            defaultValue: moment('00:00:00', 'HH:mm:ss'),
                          }}
                          format={'DD/MM/YYYY HH:mm:ss'}
                          {...input}
                        />
                      </label>
                    )}
                  </Field>
                </div>
                <div>
                  <Field name="beforeDate">
                    {({ input }) => (
                      <label className="flex flex-row items-center">
                        <span className="mr-2">
                          {t('events.filters.beforeDate')}
                        </span>
                        <DatePicker
                          showNow={true}
                          showTime={{
                            defaultValue: moment('23:59:59', 'HH:mm:ss'),
                          }}
                          format={'DD/MM/YYYY HH:mm:ss'}
                          {...input}
                        />
                      </label>
                    )}
                  </Field>
                </div>
              </div>
              <div className="flex ">
                <Field name="fields">
                  {({ input }) => (
                    <Button
                      variant="primary"
                      onClick={() => {
                        input.onChange({
                          ...input.value,
                          '': input.value[''] || '',
                        });
                        focusFieldsInput();
                      }}
                    >
                      {t('events.filters.specific.add')}
                    </Button>
                  )}
                </Field>
              </div>
            </div>
            <div className="mb-4 flex flex-row">
              <Button className="!text-xs" onClick={reset}>
                {t('events.filters.reset')}
              </Button>
              <Popover
                onOpenChange={setSuggestionsPopupState}
                open={suggestionsPopupState}
                showArrow
                placement="bottom"
                content={
                  <div className="flex flex-col">
                    {builtinFilters.map(({ label, filter }) => (
                      <Button
                        key={label}
                        onClick={() => {
                          updateValue({
                            values: {
                              fields: filter,
                            },
                          });
                          setSuggestionsPopupState(false);
                        }}
                        icon={<FilterOutlined />}
                        className="!text-left"
                      >
                        {label}
                      </Button>
                    ))}
                  </div>
                }
                trigger={['click']}
              >
                <Button className="!text-xs">
                  {t('events.filters.suggestions.title')}
                </Button>
              </Popover>
            </div>
            <Field name="fields">
              {({ input }) =>
                input.value &&
                Object.keys(input.value).length > 0 && (
                  <div
                    className="flex flex-1 pt-4 pb-8 border-t"
                    ref={fieldsInput}
                  >
                    <ObjectInput
                      {...input}
                      label={
                        <label className="-mb-2">
                          {t('events.filters.specific.label')}
                        </label>
                      }
                      removeLabel={t('events.filters.specific.remove')}
                      deleteIconClassName="!text-gray"
                    />
                  </div>
                )
              }
            </Field>
          </form>
        )}
      </Form>
    </div>
  );
};

export default Filters;
