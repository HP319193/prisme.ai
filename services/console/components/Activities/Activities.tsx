import { useTranslation } from 'react-i18next';
import { useCallback, useEffect, useState } from 'react';
import { useScrollListener } from '../useScrollListener';
import useLocalizedText from '../../utils/useLocalizedText';
import { Button, Layout, StretchContent } from '@prisme.ai/design-system';
import { Badge, PageHeader } from 'antd';
import HorizontalSeparatedNav from '../HorizontalSeparatedNav';
import Filters from './Filters';
import { useQueryString } from '../../providers/QueryStringProvider';
import { useWorkspace } from '../../providers/Workspace';
import { useEvents } from '../../providers/Events';
import EventsList from './EventsList';
import EmptyActivities from './EmptyActivities';
import { ExceptionOutlined } from '@ant-design/icons';

export const Activities = () => {
  const { t } = useTranslation('workspaces');
  const { localize } = useLocalizedText();
  const {
    workspace: { name: workspaceName },
  } = useWorkspace();
  const {
    events,
    loading,
    setFilters,
    fetchEvents,
    fetchNextEvents,
    hasMore,
    read,
    isRead,
    isVirgin,
  } = useEvents();
  const { queryString, setQueryString } = useQueryString();
  const { ref, bottom } = useScrollListener<HTMLDivElement>({ margin: -1 });

  const filtersCount = Array.from(queryString.entries()).filter(
    ([key]) => key !== 'text'
  ).length;
  const [showFilters, setShowFilters] = useState(filtersCount > 0);

  const updateQuery = useCallback(
    (text: string) => {
      setQueryString((prevQuery) => {
        const newQuery = new URLSearchParams(prevQuery);
        if (text) {
          newQuery.set('text', text);
        } else {
          newQuery.delete('text');
        }

        if (newQuery.toString() === prevQuery.toString()) return prevQuery;
        return newQuery;
      });
    },
    [setQueryString]
  );

  useEffect(() => {
    const t = setTimeout(() => {
      const query = Array.from(queryString.entries()).reduce(
        (prev, [k, v]) => ({
          ...prev,
          [k]: v,
        }),
        {}
      );
      setFilters(query);
    }, 200);

    return () => {
      clearTimeout(t);
    };
  }, [fetchEvents, queryString, setFilters]);

  useEffect(() => {
    if (bottom) {
      fetchNextEvents();
    }
  }, [bottom, fetchNextEvents]);

  const noResult = events.size === 0;

  return (
    <Layout
      Header={
        <div className="border-b  -mt-[1px]">
          <PageHeader
            className="h-[4rem] flex items-center pr-page-header-full-left !border-0"
            title={
              <HorizontalSeparatedNav className="flex-1">
                <HorizontalSeparatedNav.Separator>
                  <span className="text-base font-bold">
                    {t('events.title')}
                  </span>
                </HorizontalSeparatedNav.Separator>
                <HorizontalSeparatedNav.Separator className="flex-1">
                  <input
                    className="flex-1 focus:outline-none text-base"
                    placeholder={t('events.search.placeholder')}
                    value={queryString.get('text') || ''}
                    onChange={({ target: { value } }) => updateQuery(value)}
                  />
                </HorizontalSeparatedNav.Separator>
              </HorizontalSeparatedNav>
            }
            extra={[
              <Button
                key="filters"
                variant="grey"
                className={showFilters ? '' : '!bg-dark-accent !text-white'}
                onClick={() => setShowFilters(!showFilters)}
              >
                {filtersCount > 0 && (
                  <Badge count={filtersCount} status="warning" />
                )}
                {t('events.filters.title')}
                {filtersCount > 0 && (
                  <span className="ml-1 text-xs">({filtersCount})</span>
                )}
              </Button>,
            ]}
          ></PageHeader>
          <StretchContent visible={showFilters}>
            <Filters />
          </StretchContent>
        </div>
      }
      className="h-full"
    >
      <div className="flex h-full overflow-auto flex-col" ref={ref}>
        {!loading && isVirgin && <EmptyActivities />}
        {!loading && noResult && (
          <div className="flex flex-1 justify-center items-center">
            <div className="flex flex-1 justify-center items-center flex-col">
              <ExceptionOutlined className="text-[50px] !text-gray mb-5" />
              {t('events.filters.empty')}
            </div>
          </div>
        )}
        {!noResult && !isVirgin && (
          <>
            <EventsList
              events={events}
              workspaceName={localize(workspaceName)}
              read={read}
              isRead={isRead}
            />
            {hasMore && (
              <Button onClick={fetchNextEvents}>{t('events.more')}</Button>
            )}
          </>
        )}
      </div>
    </Layout>
  );
};
export default Activities;
