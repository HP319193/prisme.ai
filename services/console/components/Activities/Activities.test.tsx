import renderer, { act } from 'react-test-renderer';
import Activities from './Activities';
import { workspaceContext } from '../../providers/Workspace';
import { eventsContext } from '../../providers/Events';
import { queryStringContext } from '../../providers/QueryStringProvider';
import workspaceContextValue from '../../providers/Workspace/workspaceContextValue.mock';
import eventsContextValue from '../../providers/Events/eventsContextValue.mock';
import queryStringContextValue from '../../providers/QueryStringProvider/queryStringContextValue.mock';

jest.useFakeTimers();

jest.mock('../../utils/useYaml', () => {
  const toJSON = jest.fn();
  const toYaml = jest.fn();
  const useYaml = jest.fn(() => ({
    toJSON,
    toYaml,
  }));
  return useYaml;
});

jest.mock('../../layouts/WorkspaceLayout/context', () => {
  const mock = {
    createAutomation: jest.fn(),
    createPage: jest.fn(),
    installApp: jest.fn(),
  };
  return {
    useWorkspaceLayout: () => mock,
  };
});

it('should render virgin workspace', () => {
  const root = renderer.create(
    <queryStringContext.Provider value={queryStringContextValue}>
      <workspaceContext.Provider value={workspaceContextValue}>
        <eventsContext.Provider
          value={{ ...eventsContextValue, isVirgin: true, loading: false }}
        >
          <Activities />
        </eventsContext.Provider>
      </workspaceContext.Provider>
    </queryStringContext.Provider>
  );

  expect(root.toJSON()).toMatchSnapshot();
});

it('should render some events', () => {
  const root = renderer.create(
    <queryStringContext.Provider value={queryStringContextValue}>
      <workspaceContext.Provider value={workspaceContextValue}>
        <eventsContext.Provider
          value={{
            ...eventsContextValue,
            events: new Set([
              {
                id: '4',
                type: 'event a',
                createdAt: new Date('2022-01-02T08:02'),
                source: {
                  host: {
                    service: 'osef',
                  },
                  correlationId: 'osef',
                },
                size: 0,
              },
              {
                id: '3',
                type: 'event c',
                createdAt: new Date('2022-01-01T10:01'),
                source: {
                  host: {
                    service: 'osef',
                  },
                  correlationId: 'osef',
                },
                size: 0,
              },
              {
                id: '2',
                type: 'event b',
                createdAt: new Date('2022-01-01T10:00'),
                source: {
                  host: {
                    service: 'osef',
                  },
                  correlationId: 'osef',
                },
                size: 0,
              },
              {
                id: '1',
                type: 'event a',
                createdAt: new Date('2022-01-01T00:00'),
                source: {
                  host: {
                    service: 'osef',
                  },
                  correlationId: 'osef',
                },
                size: 0,
              },
            ]),
          }}
        >
          <Activities />
        </eventsContext.Provider>
      </workspaceContext.Provider>
    </queryStringContext.Provider>
  );

  expect(root.toJSON()).toMatchSnapshot();
});

it('should render no result', () => {
  const root = renderer.create(
    <queryStringContext.Provider value={queryStringContextValue}>
      <workspaceContext.Provider value={workspaceContextValue}>
        <eventsContext.Provider
          value={{
            ...eventsContextValue,
            events: new Set(),
          }}
        >
          <Activities />
        </eventsContext.Provider>
      </workspaceContext.Provider>
    </queryStringContext.Provider>
  );

  expect(root.toJSON()).toMatchSnapshot();
});

it('should update query string', () => {
  const root = renderer.create(
    <queryStringContext.Provider value={queryStringContextValue}>
      <workspaceContext.Provider value={workspaceContextValue}>
        <eventsContext.Provider value={eventsContextValue}>
          <Activities />
        </eventsContext.Provider>
      </workspaceContext.Provider>
    </queryStringContext.Provider>
  );
  const headerEl = root.root.find((item) => {
    return !!item.props.Header;
  });
  const onChange =
    headerEl.props.Header.props.children[0].props.title.props.children[1].props
      .children.props.onChange;
  act(() => {
    onChange({ target: { value: 'foo' } });
  });
  const setQueryString = queryStringContextValue.setQueryString as jest.Mock;
  expect(setQueryString).toHaveBeenCalled();
  expect(setQueryString.mock.calls[0][0](new URLSearchParams())).toEqual(
    new URLSearchParams({ text: 'foo' })
  );
  setQueryString.mockClear();

  act(() => {
    onChange({ target: { value: '' } });
  });
  expect(setQueryString).toHaveBeenCalled();
  expect(setQueryString.mock.calls[0][0](new URLSearchParams())).toEqual(
    new URLSearchParams()
  );
});
