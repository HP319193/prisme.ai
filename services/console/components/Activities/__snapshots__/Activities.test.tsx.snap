// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render no result 1`] = `
<div
  className="flex h-full overflow-auto flex-col"
>
  <div
    className="flex flex-1 justify-center items-center"
  >
    <div
      className="flex flex-1 justify-center items-center flex-col"
    >
      events.filters.empty
    </div>
  </div>
</div>
`;

exports[`should render some events 1`] = `
<div
  className="flex h-full overflow-auto flex-col"
>
  events.more
</div>
`;

exports[`should render virgin workspace 1`] = `
<div
  className="flex h-full overflow-auto flex-col"
>
  <div
    className="flex flex-1 justify-between flex-col p-28"
  >
    <div
      className="flex flex-col flex-1 justify-center w-[17rem]"
    >
      <div
        className="font-bold text-2xl my-2"
      >
        a
      </div>
      <div>
        b
      </div>
    </div>
    <div
      className="flex flex-col"
    >
      <div
        className="my-8"
      >
        workspace.empty.subtitle
      </div>
      <div
        className="flex flex-row"
      >
        <button
          className="
        focus:outline-none
        flex flex-col rounded flex-1 m-2 p-4 text-sm text-left justify-between
        transition-all group
        hover:shadow-2xl hover:-translate-y-2
        !ml-0
      "
          onClick={[MockFunction]}
          style={
            Object {
              "backgroundColor": "#F7F8D0",
            }
          }
        >
          <div
            className="flex flex-col"
          >
            <div
              className="font-bold mb-2"
            >
              workspace.empty.suggestions.page.title
            </div>
            <div
              className="relative min-h-[55px]"
            >
              <div
                className="transition-all opacity-100 group-hover:opacity-0"
              >
                workspace.empty.suggestions.page.text
              </div>
              <div
                className=" absolute top-0 flex flex-row mt-4 py-2 px-4 rounded text-3xl bg-[white] transition-all whitespace-nowrap overflow-hidden opacity-0 group-hover:opacity-100 "
                style={
                  Object {
                    "color": "#F7F8D0",
                  }
                }
              >
                +
              </div>
            </div>
          </div>
        </button>
        <button
          className="
        focus:outline-none
        flex flex-col rounded flex-1 m-2 p-4 text-sm text-left justify-between
        transition-all group
        hover:shadow-2xl hover:-translate-y-2
        undefined
      "
          onClick={[MockFunction]}
          style={
            Object {
              "backgroundColor": "#FF9261",
            }
          }
        >
          <div
            className="flex flex-col"
          >
            <div
              className="font-bold mb-2"
            >
              workspace.empty.suggestions.automation.title
            </div>
            <div
              className="relative min-h-[55px]"
            >
              <div
                className="transition-all opacity-100 group-hover:opacity-0"
              >
                workspace.empty.suggestions.automation.text
              </div>
              <div
                className=" absolute top-0 flex flex-row mt-4 py-2 px-4 rounded text-3xl bg-[white] transition-all whitespace-nowrap overflow-hidden opacity-0 group-hover:opacity-100 "
                style={
                  Object {
                    "color": "#FF9261",
                  }
                }
              >
                +
              </div>
            </div>
          </div>
        </button>
        <button
          className="
        focus:outline-none
        flex flex-col rounded flex-1 m-2 p-4 text-sm text-left justify-between
        transition-all group
        hover:shadow-2xl hover:-translate-y-2
        undefined
      "
          onClick={[MockFunction]}
          style={
            Object {
              "backgroundColor": "#E7F6F6",
            }
          }
        >
          <div
            className="flex flex-col"
          >
            <div
              className="font-bold mb-2"
            >
              workspace.empty.suggestions.app.title
            </div>
            <div
              className="relative min-h-[55px]"
            >
              <div
                className="transition-all opacity-100 group-hover:opacity-0"
              >
                workspace.empty.suggestions.app.text
              </div>
              <div
                className=" absolute top-0 flex flex-row mt-4 py-2 px-4 rounded text-3xl bg-[white] transition-all whitespace-nowrap overflow-hidden opacity-0 group-hover:opacity-100 "
                style={
                  Object {
                    "color": "#E7F6F6",
                  }
                }
              >
                +
              </div>
            </div>
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;
